### PR TITLE
Optimize performance of softmax_bwd when axis!=-1

### PR DESCRIPTION
**Pull request by ZzSean**
### PR types

Performance optimization
### PR changes

OPs
### Describe

Optimize performance of softmax_bwd when axis!=-1

优化思路跟reduce类似，主要是两个：
1）当规约的维度不是最低维时，让每个warp还可以保证合并访存；
2）使用二维的 block 和 grid，尽可能地保证 thread 的并行度和资源利用最大化

| case                         | pytorch | paddle 优化前 | diff          | paddle 优化后 | diff          | 加速比 |
| ---------------------------- | ------- | ------------- | ------------- | ------------- | ------------- | ------ |
| fp32,[128,128,16,16]  axis:0 | 0.17740 | 0.12211       | 优于 (45.28%) | 0.08359       | 优于 (1.12x)  | 1.46   |
| fp16,[128,128,16,16] axis:0  | 0.15025 | 0.10154       | 优于 (47.97%) | 0.06241       | 优于 (1.41x)  | 1.63   |
| fp32,[512,896,4,12] axis:1   | 1.01407 | 0.90445       | 优于 (12.12%) | 0.61183       | 优于 (65.74%) | 1.48   |
| fp16,[512,896,4,12] axis:1   | 0.68431 | 1.03145       | 差于 (50.73%) | 0.35872       | 优于 (90.76%) | 2.88   |

### CODE REVIEW

**Review by AshburnLee — COMMENTED** [review]
LGTM

**Review by limin2021 — APPROVED** [review]
LGTM.
